// @splunk/otel/lib/index.js
"use strict";
/*
 * Copyright Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.stopTracing = exports.startTracing = exports.startProfiling = exports.startMetrics = exports.listEnvVars = exports.stop = exports.start = exports.ConsoleMetricExporter = void 0;
const util_1 = require("util");
const tracing_1 = require("./tracing");
const metrics_1 = require("./metrics");
var ConsoleMetricExporter_1 = require("./metrics/ConsoleMetricExporter");
Object.defineProperty(exports, "ConsoleMetricExporter", { enumerable: true, get: function () { return ConsoleMetricExporter_1.ConsoleMetricExporter; } });
const profiling_1 = require("./profiling");
var start_1 = require("./start");
Object.defineProperty(exports, "start", { enumerable: true, get: function () { return start_1.start; } });
Object.defineProperty(exports, "stop", { enumerable: true, get: function () { return start_1.stop; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "listEnvVars", { enumerable: true, get: function () { return utils_1.listEnvVars; } });
exports.startMetrics = (0, util_1.deprecate)(metrics_1.startMetrics, 'startMetrics is deprecated. Use generic start() and stop() functions instead');
exports.startProfiling = (0, util_1.deprecate)(profiling_1.startProfiling, 'startProfiling is deprecated. Use generic start() and stop() functions instead');
exports.startTracing = (0, util_1.deprecate)(tracing_1.startTracing, 'startTracing is deprecated. Use generic start() and stop() functions instead');
exports.stopTracing = (0, util_1.deprecate)(tracing_1.stopTracing, 'stopTracing is deprecated. Use generic start() and stop() functions instead');
